### DISCUSSION TITLE
how do I calculate data on collection of game posts

### DISCUSSION THREAD

**Discussion by btrem** [11ty/eleventy#3180] [Q&A] [answered]
I have a collection of game result posts:
```
eleventyConfig.addCollection('gamePosts', function(collectionApi) {
    return collectionApi.getFilteredByGlob('src/posts/games/**/*.md');
});
```
I loop through the collection and show them on a page. I also show game statistics (e.g., win percentage, winning streak, etc.) on that same page. What I have now is a new collection:
```
eleventyConfig.addCollection('gameStats', function(collectionApi) {
    let gamePosts = collectionApi.getFilteredByGlob('src/posts/games/**/*.md');

    // loop through collection, calculate stats
    // and store in variables, e.g., `streak`, `winPercentage`

    return {
        streak: streak,
        winPercentage: winPercentage
        // there's additional stats, removed here to keep example simple
    };

});
```
So I'm recreating the collection, which feels wrong. It seems like filter is the more appropriate tool, where I could pass my game post collection to a filter and have it return stats. But then I'd need a filter for each stat, like `eleventyConfig.addFilter('gameStreak', function(myGameCollection) /* calculate winning streak */ })` and `eleventyConfig.addFilter('winPercent', function(myGameCollection) /* calculate winning percent */ })`. But if I go this route, each filter would make 11ty loop over the collection, instead of doing it once and getting all the results.

So what's the best way to calculate data on a collection?

**Comment by pdehaan** [comment, marked as answer]
Not sure I'm 100% understanding your setup, but you could create a single filter that takes any arbitrary collection as an argument and returns a "stats" object (which sounds easier than creating _n_ filters; one per stat).

```js
  eleventyConfig.addFilter("gameStats", function (collection = []) {
    return collection.reduce((acc, game) => {
      const { points } = game.data;
      acc.winPercentage = points[0] + points[1];
      acc.streak = acc.streak + 1;
      return acc;
    }, { streak: 0 });
  });
```

Where I created some fake data with front matter like this:
```liquid
---
title: Golden State Warriors at Utah Jazz
points:
  - 88
  - 85
winner: Golden State Warriors
---

<h1>{{ title }}</h1>
```

So verrrry terrible stats calculations, but yeah... I also created a src/posts/games/games.11tydata.json which sets `tags` for all child templates in that directory:
```json
{
  "tags": ["games"]
}
```

Now I can just refer to `collections.games` instead of needing something like `collectionApi.getFilteredByGlob('src/posts/games/**/*.md')`.

And my very bad, no good src/index.liquid looks something like this:

```liquid
---
title: Stats
---
=
{%- assign stats = collections.games | gameStats -%}

<h1>{{ title }}</h1>
<pre>{{ stats | json }}</pre>

<p>Win Streak: {{ stats.streak }}</p>
<p>Win Pct%: {{ stat.winPercentage }}</p>
```

And the generated HTML looks like:
```html
<h1>Stats</h1>
<pre>{"streak":2,"winPercentage":173}</pre>

<p>Win Streak: 2</p>
<p>Win Pct%: 173</p>
```

I think one benefit of this is if you have multiple different seasons of games, you could just add new `tags` in the right subfolders and then loop over "collections.games_23_24 | gameStats" to loop over a subset of games.

**Reply by btrem** [reply]
Using `assign` was the part I was missing (well, `set`, because I use nunjucks). I tend to avoid `set` because it seems like a recipe for putting logic in templates, but in this case it is the least worst of the three options.

And good idea about subfolders. I'm not ready to tackle that yet, but it might be fun to calculate stat changes over time.

Thanks as always for your help.
